### PR TITLE
Add OpInfo-based tests for elementwise multiply on sparse COO/CSR/CSC/BSR/BSC tensors

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

In addition, the PR enables half and half-complex types support in sparse COO multiply.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98288
* #98292
* #100394
* #100393
* #100392
* #100391
* #100389

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -9246,7 +9250,23 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
...
                     supports_forward_ad=True,
                     supports_fwgrad_bwgrad=True,
-                    supports_two_python_scalars=True),
+                    supports_two_python_scalars=True,
+                    supports_sparse=True,
+                    supports_sparse_csr=True,
+                    supports_sparse_csc=True,
+                    supports_sparse_bsr=True,
+                    supports_sparse_bsc=True,
+                    sample_inputs_sparse_coo_func=partial(sample_inputs_elementwise_binary_operation_sparse,
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:9259]: Only related to this PR: it seems the logic for defining support for sparse and setting the corresponding sample inputs function is a little redundant. Meaning, you need to set `supports_sparse=True` *and* pass a `sample_inputs_sparse_coo_func`. Shouldn't the latter be enough?

Internally one can use `self.sample_inputs_sparse_coo_func is not None` over `self.supports_sparse`. For BC, one could also add `self.supports_sparse = self.sample_inputs_sparse_coo_func is not None` to the constructor.

File: test/test_sparse.py [hunk 2]
```diff
@@ -4761,6 +4768,88 @@ def test_gradcheck_mm(self, layout, dtype, device, masked, fast_mode):
...
+                except Exception as msg:
+                    # this is strided op issue, so skipping the sample silently here
+                    if "\"cpublas_axpy_impl\" not implemented for 'ComplexHalf'" in str(msg):
+                        continue
+                    raise
+                self.assertEqual(result, dense)
+
+        if count == 0:
+            # we count samples to avoid false-positive test reports
+            self.skipTest('no sample inputs')
```
Comment (pmeier) [inline test/test_sparse.py:4850]: Shouldn't we fail here since that seems like a setup issue? Skips often get ignored because most people just look for a boolean signal from CI.

File: test/test_sparse.py [hunk 3]
```diff
@@ -4761,6 +4768,88 @@ def test_gradcheck_mm(self, layout, dtype, device, masked, fast_mode):
...
+    @suppress_warnings
+    @ops(binary_ufuncs_with_sparse_support)
+    @all_sparse_layouts('layout', include_strided=False)
+    def test_binary_operation(self, layout, device, dtype, op):
+        count = 0
+        for sample in op.sample_inputs_sparse(layout, device, dtype):
+            count += 1
+            t_inp, t_args, t_kwargs = sample.input, sample.args, sample.kwargs
+            batch_dim = t_inp.dim() - t_inp.dense_dim() - t_inp.sparse_dim()
+            # TODO: eliminate the if-blocks below
```
Comment (pmeier) [inline test/test_sparse.py:4781]: TBH, it is really hard to parse what is actually happening below. IMO in a test that parametrizes over `@ops(binary_ufuncs_with_sparse_support)` I do not expect special cases for individual operators like `if op.name == 'mul':`. This test should only handle the generic test, which seems to be `self.assertEqual(result, dense)`.

For operator specific assertions, we should have a separate test like

```
def test_binary_operation_mul(self, ...):
    op = next(op for op in binary_ufuncs_with_sparse_support if op.name == "mul")
```

**Review by pmeier — COMMENTED** [review]
I reviewed the testing part and defer the review of sparse to someone else.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -19259,6 +19278,15 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
     PythonRefInfo(
         "_refs.addcmul",
         torch_opinfo_name="addcmul",
+        skips=(
+            # Reference result was farther (1.3343989849090576e-05)
+            # from the precise computation than the torch result
+            # was (9.592622518539429e-06)!
+            DecorateInfo(unittest.skip("Unstable expected failure. Skipped!"), 'TestCommon', 'test_python_ref',
```
[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:19285]: yes, it is flaky: some CI jobs fail while some others don't.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:9259]: IIUC, while `supports_sparse=True` and defining `sample_inputs_sparse_coo_func` have overlapping meanings, then just specifying `supports_sparse=True` makes also sense: in principle some sparse tests (say, conversion tests) may use strided samples.

Otherwise, I agree, defining `sample_inputs_sparse_coo_func` could imply `supports_sparse=True`.

Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:9259]: Fair enough. Maybe an improvement could be to have `supports_sparse: Optional[bool] = None` as parameter and do

```python
if supports_sparse is None:
    supports_sparse = sample_inputs_sparse_coo_func is not None
```

This way one can set `supports_sparse=True` if needs be, but otherwise it will be implied by the presence of a sample inputs function.

File: test/test_sparse.py (hunk 2 above)
Comment (pearu) [reply test/test_sparse.py:4850]: Yes, we can fail here.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:19285]: Correct me if I'm wrong here, but the purpose of this test is to check that the Python reference implementation is doing the same thing as the C++ implementation, right? If so, isn't the exact result irrelevant here? It seems that the correctness of the C++ op is checked elsewhere.

If that is the case, I don't think we should skip here, but rather increase the tolerances to avoid the flakiness. Otherwise the two implementations can drift apart and we would never know. Granted, with tolerances set they can also drift a little, but that is a lot more safety than a blanket skip.

[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py:19285]: Yes, you are right. However, changing the tolerances is not implemented in the python_ref tests:
https://github.com/pytorch/pytorch/blob/752e43c301ac433b909c21e2b34bc3428afbb173/test/test_ops.py#L441
and it would be OT for this PR.

[the remark this answers is not in the document]
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:19285]: That is unfortunate. In that case could you change the comment to a `FIXME` explaining that we should set tolerances there, but the test doesn't support it?

[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py]: Added FIXME comment.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py]: Done: for unspecified `supports_sparse`, specifying `sample_inputs_sparse_coo_func` now implies `supports_sparse=True`

File: test/test_sparse.py (hunk 2 above)
Comment (pearu) [reply test/test_sparse.py]: Fixed: replaced `count`-related code with the `non_empty_sampler` wrapper in `OpInfo.sample_inputs_sparse` method that will raise an exception when a sample generator does not produce any samples.

File: test/test_sparse.py (hunk 3 above)
Comment (pearu) [reply test/test_sparse.py]: I have updated the TODO comment to indicate that the if-blocks are due to unimplemented features and are temporary.

**Review by pmeier — APPROVED** [review]
Re https://github.com/pytorch/pytorch/pull/98288#discussion_r1158220108:

> there are no operator-specific tests here other than the failures that ought to be removed eventually

I'm still seeing a lot of `if op.name == 'mul'`. Doesn't that mean that we have "operate-specific tests" here? IMO they should be factored out in a separate test. More so, these samples should probably removed from the sample generation for now if they lead to errors, i.e. the test should not be responsible to filter out failing samples that are specific to one operator.

Anyway, not going to block over this. Thanks for addressing the rest of my comments!

**Comment by pearu** [comment]
> Re [#98288 (comment)](https://github.com/pytorch/pytorch/pull/98288#discussion_r1158220108):
> 
> > there are no operator-specific tests here other than the failures that ought to be removed eventually
> 
> I'm still seeing a lot of `if op.name == 'mul'`. Doesn't that mean that we have "operate-specific tests" here? IMO they should be factored out in a separate test. More so, these samples should probably removed from the sample generation for now if they lead to errors,
> i.e. the test should not be responsible to filter out failing samples that are specific to one operator.

Atm, the test is used only for `mul` operation, however, it is very likely that the same sample will fail for other binary operations as well as when the origin of failures is the same for different ops.

> Anyway, not going to block over this. Thanks for addressing the rest of my comments!

Thanks!

**Comment by pmeier** [comment]
> it is very likely that the same sample will fail for other binary operations as well

Isn't that a strong argument to remove this sample from the generation in general? Why generate it if it always fails?

**Comment by pearu** [comment]
> > it is very likely that the same sample will fail for other binary operations as well
> 
> Isn't that a strong argument to remove this sample from the generation in general? Why generate it if it always fails?

No. Eventually, we want to fix all these failures and it is not the sample generator's fault that some of its samples are failing on certain operations while for other ops the samples are ok.

**Comment by pmeier** [comment]
> it is not the sample generator's fault that some of its samples are failing on certain operations while for other ops the samples are ok.

True, but in that case you are using the wrong sampler. Instead of using `sample_inputs_elementwise_binary_operation_sparse` directly, it should be a separate `sample_inputs_mul_sparse`, which can call the general sampler, but needs to filter out broken samples.

**Comment by pearu** [comment]
> True, but in that case you are using the wrong sampler. Instead of using sample_inputs_elementwise_binary_operation_sparse directly, it should be a separate sample_inputs_mul_sparse, which can call the general sampler, but needs to filter out broken samples.

This filtering approach is appropriate for cases where certain samples are invalid as inputs for a particular operation (say, rhs to division cannot contain zeros, for instance) but in this case here, the operation is broken, not the samples.

In general, the variability of sparse layouts, operation parameters, and what is implemented/not implemented is much greater than in the case of strided layout and I think it would be impractical to create so many different samplers according to what operations on sparse tensors currently support or not.

**Comment by pmeier** [comment]
> This filtering approach is appropriate for cases where certain samples are invalid as inputs for a particular operation (say, rhs to division cannot contain zeros, for instance) but in this case here, the operation is broken, not the samples.

The filtering has to happen somewhere. Right now it happens in the test, which creates the `if op.name == "mul"` conditions. All I'm saying is to do the same, but rather in the sample generation. This would keep the test clean and easier to comprehend. 

If you want to make sure that these samples are later re-added when the op is fixed, create a second test with the "broken" inputs and xfail the test. As soon as the test passes unexpectedly, you can remove it and the filtering in the generation as well.

Anyway, I'm going to disengage here as I'm not the person maintaining the test in question.

**Comment by pearu** [comment]
> The filtering has to happen somewhere. Right now it happens in the test, which creates the `if op.name == "mul"` conditions. All I'm saying is to do the same, but rather in the sample generation. This would keep the test clean and easier to comprehend.
> 
> If you want to make sure that these samples are later re-added when the op is fixed, create a second test with the "broken" inputs and xfail the test. As soon as the test passes unexpectedly, you can remove it and the filtering in the generation as well.

I have moved the logic of filtering failing/succeeding samples into sample_inputs functions so that usage of `op.name` in tests is minimized.

> Anyway, I'm going to disengage here as I'm not the person maintaining the test in question.

@pmeier Thanks for the reviews and discussions so far, it has been very helpful! Since I have reorganized the testing logic, could you give a review once more?

**Review by pmeier — COMMENTED** [review]
I think using `ErrorInput`'s is the right direction, since it aligns this with the rest of the test suite. However, I don't understand why you want to mix them with the regular `SampleInput`'s. I see three downsides with this:

1. As developer, I don't expect `op.sample_inputs_sparse` to return `ErrorInput`'s. Since they behave differently from `SampleInput`'s this makes it harder to understand what is going on.
2. Even if one understands 1., we have to filter them out in every single test. Right now, there are only two, but I guess there will be more in the future?
3. Even if we are ok with filtering them out every time, if the generation is expensive (I don't know if that is the case), we are doing it multiple times for no reason.

My suggestion here is to just drop them completely from the sample inputs function and have a separate `error_inputs_func` that generates them. This way, the error inputs will be tested by the regular framework as expected and the sparse tests don't need a new paradigm that a sample might be an `ErrorInput` as well.

If you really want to avoid code duplication, here is my suggestion on how to set this up:

```python
def _sample_error_inputs_mul_sparse(op_info, device, dtype, requires_grad, layout, **kwargs):
    for sample in sample_inputs_elementwise_binary_operation_sparse(
        op_info, device, dtype, requires_grad, layout, **kwargs
    ):
        t_inp, t_args, t_kwargs = sample.input, sample.args, sample.kwargs
        batch_dim = t_inp.dim() - t_inp.dense_dim() - t_inp.sparse_dim()
        if layout is torch.sparse_csr and batch_dim > 0 and t_args[0].ndim > 0:
            yield ErrorInput(
                sample,
                error_regex="crow_indices is supposed to be a vector, but got 2 dimensional tensor",
            )
        elif ...:
            yield ErrorInput(...)
        else:
            yield sample

def sample_inputs_mul_sparse(op_info, device, dtype, requires_grad, layout, **kwargs):
    return filter(
        lambda sample: isinstance(sample, SampleInput),
        _sample_error_inputs_mul_sparse(op_info, device, dtype, requires_grad, layout),
    )

def error_inputs_mul(op_info, device, dtype, requires_grad, **kwargs):
    # I didn't see one, but if there already is an error inputs function, this can be merged into it
    for layout in [...]:
        for sample in _sample_error_inputs_mul_sparse(op_info, device, dtype, requires_grad, layout):
            if isinstance(sample, ErrorInput):
                yield sample
```

Consider this review is non-blocking.

File: torch/testing/_internal/opinfo/definitions/sparse.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+
+def sample_inputs_mul_sparse(op_info, device, dtype, requires_grad, layout, **kwargs):
+    """Sample inputs for mul operation on sparse tensors."""
+    for sample in sample_inputs_elementwise_binary_operation_sparse(
+        op_info, device, dtype, requires_grad, layout, **kwargs
+    ):
+        if isinstance(sample, ErrorInput):
+            yield sample
+            continue
```
[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/opinfo/definitions/sparse.py:88]: Right. It will be needed only when drawing sample input functions that may yield ErrorInptu instances.

File: test/test_sparse.py [hunk 6]
```diff
@@ -4758,6 +4780,54 @@ def test_gradcheck_mm(self, layout, dtype, device, masked, fast_mode):
...
+        if not op.supports_sparse_layout(layout):
+            self.skipTest(f'{layout} is not supported in `{op.name}` OpInfo definition. Skipping!')
+
+        for sample in op.sample_inputs_sparse(layout, device, dtype):
+
+            if isinstance(sample, ErrorInput):
+                with self.assertRaisesRegex(sample.error_type, sample.error_regex):
+                    t_inp, t_args, t_kwargs = sample.sample_input.input, sample.sample_input.args, sample.sample_input.kwargs
+                    op.op(t_inp, *t_args, **t_kwargs)  # ######### WHEN SUCCEEDS, UPDATE opinfo/definitions/sparse.py
+                continue
```
[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse.py:4797]: That looks perfect. Thanks for the hint!

File: torch/testing/_internal/opinfo/definitions/sparse.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        t_inp, t_args, t_kwargs = sample.input, sample.args, sample.kwargs
+        batch_dim = t_inp.dim() - t_inp.dense_dim() - t_inp.sparse_dim()
+        if layout is torch.sparse_csr and batch_dim > 0 and t_args[0].ndim > 0:
+            yield ErrorInput(
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torch/testing/_internal/opinfo/definitions/sparse.py:92]: https://github.com/pytorch/pytorch/blob/d44c5e3639c8d27d454db931c15ce28794697a8d/torch/testing/_internal/opinfo/core.py#L320

File: test/test_sparse.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse.py:4797]: There are a couple of downsides to using `test_errors` from test_ops.py:
- it has a considerable overhead of importing tests
- `error_input_func` does not support `dtype` and `requires_grad` arguments which means that the error functions need to have a loop over these arguments rather than specifying these via decorators.

[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_sparse.py:4797]: > it has a considerable overhead of importing tests

I don't understand. What are you importing? There is no need to import anything. `test/test_ops.py` is run by the regular CI. By adding error inputs to the OpInfo, you get this for "free" and don't have to deal with this in the sparse tests at all.

> * `error_input_func` does not support `dtype` and `requires_grad` arguments which means that the error functions need to have a loop over these arguments rather than specifying these via decorators.

Fair enough. In the solution I sketched above, I also needed to iterate over the different layouts. Doing the same for `dtype` and `requires_grad` isn't too bad. That being said, do we actually need the whole matrix of inputs when checking for the errors? Looking at the currently implemented error inputs, I mostly see conditions based on the shape and a few that include `torch.bool`. `requires_grad` seems to be unused for the error inputs. Maybe just iterate over `dtypes = [torch.bool, torch.int32, torch.float32, torch.complex64]` or the like here?

[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse.py:4797]: An OpInfo instance has `dtypes` attribute that `error_input_func` can use. So, it's not a big deal.

[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse.py]: Now `error_input_func` is being used.

[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse.py]: A follow-up issue: https://github.com/pytorch/pytorch/issues/99404

File: torch/testing/_internal/opinfo/definitions/sparse.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/opinfo/definitions/sparse.py]: Removed.

**Review by pmeier — APPROVED** [review]
Thanks a lot @pearu for the patience. IMO, it is a lot cleaner now. My approval is only for the testing side. For meaty parts, someone with sparse knowledge needs to take over.

File: torch/testing/_internal/opinfo/definitions/sparse.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+def _filter_error_inputs(sample_and_error_inputs_func, op_info, device, **kwargs):
+    for layout in (
+        torch.sparse_csr,
+        torch.sparse_csc,
+        torch.sparse_bsr,
+        torch.sparse_bsc,
+        torch.sparse_coo,
+    ):
+        for dtype in op_info.supported_dtypes(device):
```
[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/opinfo/definitions/sparse.py:35]: Applied.

**Comment by pearu** [comment]
An observation:

The various test failures (in test/functorch/) that appear when using the existing OpInfo-based testing framework (e.g. `error_input_func`) for sparse samples illustrate a general problem with the sparse tensor testing support in PyTorch: the testing framework may use features that are available for strided tensors only (say, the usage of `repeat()` method on sample inputs as in test/functorch/common_utils.py) and when sparse sample inputs are generated, the testing framework will break down.

While there exist various ways to disable samples that break tests or operations (`xfail`, DecorateInfo skips, etc), there exists no standard way to handle the testing framework failures caused by sparse samples (that otherwise would be valid samples for operations or tests). I guess the reason for this is the assumption that the testing framework does not depend on the generated samples but unfortunately, this is not always the case. There exist no tests on the testing framework to verify this assumption.

File: test/test_sparse.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse.py]: I have introduced `error_inputs_sparse_func` to `OpInfo` as the interference with other unrelated tests when using `error_inputs_func` is unreasonably laborious.

File: aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp [hunk 9]
```diff
@@ -32,7 +32,8 @@ static void addcmul_cpu_kernel(TensorIteratorBase& iter, const Scalar& value) {
           return convert_float_bfloat16(self_vec0, self_vec1);
         });
   } else {
-    AT_DISPATCH_ALL_TYPES_AND_COMPLEX(dtype, "addcmul_cpu_out", [&] {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(at::ScalarType::ComplexHalf, at::ScalarType::Half,
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:35]: I wonder why this wasn't already done earlier.
